src/api/entities/CorporateBallot/index.ts

```diff
@@ -0,0 +1,298 @@
+   */
+  public async exists(): Promise<boolean> {
+    const {
+      id,
+      asset,
+      context: {
+        polymeshApi: { query },
+      },
+      context,
+    } = this;
+
+    const caId = corporateActionIdentifierToCaId({ localId: id, asset }, context);
+    const meta = await query.corporateBallot.metas(caId);
+
+    return meta.isSome;
+  }
+
+  /**
+   * Retrieve details associated with this Ballot
+   *
+   * @throws if the Ballot does not exist
+   */
+  public async details(): Promise<CorporateBallotDetails> {
+    const { id, asset, context } = this;
+
+    const details = await getCorporateBallotDetailsOrThrow(asset, id, context);
+
+    return details;
+  }
+
+  /**
+   * Return the Corporate Ballot's static data
+   */
+  public override toHuman(): HumanReadable {
+    const { id, asset } = this;
+
+    return toHumanReadable({
+      id,
+      assetId: asset.id,
+    });
+  }
+
+  /**
+   * Return the status of the Ballot
+   *
+   * @throws if the Ballot does not exist
+   */
+  public async status(): Promise<CorporateBallotStatus> {
+    const details = await this.details();
+
+    return ballotDetailsToBallotStatus(details);
+  }
+
+  /**
+   * Retrieve the results of the Ballot
+   *
+   * @throws if the Ballot does not exist
+   */
+  public async results(): Promise<CorporateBallotMetaWithResults> {
+    const {
+      id,
+      asset,
+      context,
+      context: {
+        polymeshApi: { query },
+      },
+    } = this;
+
+    const {
+      meta: { title: ballotTitle, motions },
+    } = await this.details();
+
+    const caId = corporateActionIdentifierToCaId({ localId: id, asset }, context);
+
+    const rawResults = await query.corporateBallot.results(caId);
+    const results = rawResults.map(result => u128ToBigNumber(result));
+
+    const metaWithResults: CorporateBallotMetaWithResults = {
+      title: ballotTitle,
+      motions: [],
+    };
+
+    let resultIndex = 0;
+
+    motions.forEach(({ title, infoLink, choices }) => {
+      const motionWithResults: CorporateBallotMotionWithResults = {
+        title,
+        infoLink,
+        choices: [],
+        total: new BigNumber(0),
+      };
+
+      choices.forEach(choice => {
+        motionWithResults.choices.push({
+          choice,
+          votes: results[resultIndex],
+        });
+        motionWithResults.total = motionWithResults.total.plus(results[resultIndex]);
+        resultIndex += 1;
+      });
+
+      metaWithResults.motions.push(motionWithResults);
+    });
+
+    return metaWithResults;
+  }
+
+  /**
+   * Retrieve the participation of the Ballot
+   *
+   * @throws if the Ballot does not exist
+   */
+  public async votesByIdentity(did: Identity | string): Promise<CorporateBallotWithParticipation> {
+    const {
+      id,
+      asset,
+      context,
+      context: {
+        polymeshApi: { query },
+      },
+    } = this;
+
+    const {
+      meta: { title: ballotTitle, motions },
+    } = await this.details();
```

Comment:
Again, calling the `details` method everytime doesn't feel right. You can extract out the `metas` part and use in these type of methods